api/v1alpha1/appgroup_types.go

```diff
@@ -314,8 +304,22 @@ func (in *ChartStatus) GetStatusConditions() *[]metav1.Condition {
 	return &in.Conditions
 }
 
+func (in *ApplicationGroup) GetLastSuccessful() *ApplicationGroupSpec {
+	lastSuccessful := &ApplicationGroupSpec{}
+	if s, ok := in.Annotations[LastSuccessfulAnnotation]; ok {
+		_ = json.Unmarshal([]byte(s), lastSuccessful)
+		return lastSuccessful
+	}
+	return nil
+}
+
+func (in *ApplicationGroup) SetLastSuccessful() {
+	b, _ := json.Marshal(&in.Spec)
+	in.SetAnnotations(map[string]string{LastSuccessfulAnnotation: string(b)})
```

Comment:
As a possible side-effect, this could delete/unset other annotations (if any).

Reply:
Good call, this is actually an issue in the other PR so changing it there

Reply:
Also, if you could please review #267 first, I want to get those tests in before making big changes to refactor the logic